includes/class-micropub-authorize.php

```diff
@@ -0,0 +1,218 @@
+	 * stripped out by the server or PHP. Typically this is then used to
+	 * generate `PHP_AUTH_USER`/`PHP_AUTH_PASS` but not passed on. We use
+	 * `getallheaders` here to try and grab it out instead.
+	 *
+	 * @return string|null Authorization header if set, null otherwise
+	 */
+	public static function get_authorization_header() {
+		if ( ! empty( $_SERVER['HTTP_AUTHORIZATION'] ) ) {
+			return wp_unslash( $_SERVER['HTTP_AUTHORIZATION'] );
+		}
+		// When Apache speaks via FastCGI with PHP, then the authorization header is often available as REDIRECT_HTTP_AUTHORIZATION.
+		if ( ! empty( $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] ) ) {
+			return wp_unslash( $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] );
+		}
+		$headers = getallheaders();
+		// Check for the authorization header case-insensitively
+		foreach ( $headers as $key => $value ) {
+			if ( strtolower( $key ) === 'authorization' ) {
+				return $value;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Validate the access token at the token endpoint.
+	 *
+	 * https://indieauth.spec.indieweb.org/#access-token-verification
+	 */
+	public static function determine_current_user( $user_id ) {
+		$cls = get_called_class();
+		// Do not try to find a user if one has already been found
+		if ( ! empty( $user_id ) ) {
+			return $user_id;
+		}
+
+		// Only try to authenticate if this is the Micropub endpoint
+		if ( ! empty( get_query_var( 'micropub' ) ) ) {
+			return $user_id;
+		}
+
+		// find the access token
+		$auth  = static::get_authorization_header();
+		$token = self::get( $_POST, 'access_token' );
+		if ( ! $auth && ! $token ) {
+			static::authorize_error( 401, 'missing access token' );
+		}
+
+		$resp = wp_remote_get(
+			get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ), array(
+				'headers' => array(
+					'Accept'        => 'application/json',
+					'Authorization' => $auth ?: 'Bearer ' . $token,
+				),
+			)
+		);
+		if ( is_wp_error( $resp ) ) {
+			return $resp;
+		}
+
+		$code           = wp_remote_retrieve_response_code( $resp );
+		$body           = wp_remote_retrieve_body( $resp );
+		$params         = json_decode( $body, true );
+		static::$scopes = explode( ' ', $params['scope'] );
+
+		if ( (int) ( $code / 100 ) !== 2 ) {
+			return static::authorize_error(
+				$code, 'invalid access token: ' . $body
+			);
+		} elseif ( empty( static::$scopes ) ) {
+			return static::authorize_error(
+				401, 'access token is missing scope'
+			);
+		}
+
+		$me                             = untrailingslashit( $params['me'] );
+		static::$micropub_auth_response = $params;
+
+		// look for a user with the same url as the token's `me` value.
+		$user = static::user_url( $me );
+
+		add_filter( 'indieauth_scopes', array( $cls, 'indieauth_scopes' ), 11 );
+		add_filter( 'indieauth_response', array( $cls, 'indieauth_response' ), 11 );
+		if ( $user ) {
+			return $user;
+		}
+
+		// no user with that url. if the token is for this site itself, allow it and
```

Comment:
in the detailed PR comment (thank you for that btw!), you said:

> As one change, if it cannot determine a user ID, it will not let the post be made, as it cannot match it up with an existing user.

...but this seems to still allow posts without a user...?

Reply:
get_current_user_id should return 0. This returns null. The reason for this difference is to allow better error handling. It also optionally returns an error function.

But, if you look at where it returns, if ! $user_id then it returns unauthorized

Reply:
huh. so if i understand you right, this block is depending on the fact that `$user_id` here is 0 and `$user` is null? if so, ok, but then change the `return`s below to `return 0` and `return null`, or whatever, explicitly. _much_ more readable.